src/future/mod.rs

```diff
@@ -128,6 +139,10 @@ where
                     }
                 });
```

Comment:
I think this "Unblock our waiter if we have one and it's still alive" code can just be deleted, along with the `set_waiter` call in this module, right? That code is just wrong -- we should be invoking the waker for whoever's waiting for us, not this weird special-case thing. I think we just copied and pasted it from the `thread` implementation, which doesn't have access to a waker and so has to build its own waker-esque mechanism with `set_waiter`/`take_waiter`.